reciprocal_cache.go

```diff
@@ -0,0 +1,134 @@
+// uint256: Fixed size 256-bit math library
+// Copyright 2021 uint256 Authors
+// SPDX-License-Identifier: BSD-3-Clause
+
+package uint256
+
+import (
+	"sync"
+)
+
+// Cache for reciprocal()
+//
+// Each cache set contains its own mutex to reduce lock contention in highly
+// multithreaded settings.
+//
+// Adjust cacheIndexBits and cacheWays to scale the size of the cache.
+// Total cache size is (24+72*cacheWays)*2^cacheIndexBits bytes, which is
+// 96 KiB with cacheIndexBits = 8 and cacheWays = 5. There are also 16 bytes
+// for hit and miss counters.
+//
+// Reasonable values are quite small, e.g. cacheIndexBits from 2 to 10, and
+// cacheWays around 5. Note that cacheWays = 5+8n makes the set size an integer
+// number of 64-byte cachelines.
+//
+// If you want to disable the cache, set cacheIndexBits and cacheWays to 0
+//
+// If you want to have a hardcoded modulus with precomputation, set
+// fixedModulus = true and adjust the value stored in fixed_m. This can be
+// used with or without the regular cache.
+
+const (
+	cacheIndexBits = 8
+	cacheWays      = 5
+
+	cacheSets = 1 << cacheIndexBits
+	cacheMask = cacheSets - 1
+
+	fixedModulus = true
+)
+
+type cacheSet struct {
+	rw  sync.RWMutex
+	mod [cacheWays]Int
+	inv [cacheWays][5]uint64
+}
+
+type reciprocalCache struct {
+	set  [cacheSets]cacheSet
+	hit  uint64
+	miss uint64
+}
+
+// NewCache returns a new reciprocalCache.
+func NewCache() *reciprocalCache {
+	return &reciprocalCache{}
+}
+
+func (c *reciprocalCache) Stats() (hit, miss uint64) {
+	return c.hit, c.miss
+}
+
+var (
+	// Fixed modulus and its reciprocal
+	fixed_m Int
+	fixed_r [5]uint64
+)
+
+func init() {
+	if fixedModulus {
+		// Initialise fixed modulus
+		fixed_m[3] = 0xffffffff00000001
+		fixed_m[2] = 0x0000000000000000
+		fixed_m[1] = 0x00000000ffffffff
+		fixed_m[0] = 0xffffffffffffffff
+
+		fixed_r = reciprocal(fixed_m, nil)
+	}
```

Comment:
This is a bit magic to me. What is that fixed modulus? Please add a comment about what it represents. 
Also, this is another case of where, if someone wants to use a different fixed modulus, they have to copy-paste the entire library and cannot use it as a go module. 

I think it may be better to define certain fixed modulus's, and hten let the caller either use one of those or supply one himself. 
Like
```
var (
	// Fixed modulus and its reciprocal
	FixedModulusCurveFoo , _ = FromHex("0xffffffff00000001000000000000000000000000ffffffffffffffffffffffff")
	FixedReciprocalCurveFoo [5]uint64
)

func init() {
	FixedReciprocalCurveFoo = reciprocal(*FixedModulusCurveFoo, nil)
}
```
And then the cache itself could be initalized with a the fixed modulus to use, if any

Reply:
Pushed a commit to address this, consider it a suggestion.